Add database to monolith

mcruzdev opened the pull request:
I replaced the in memory storage like a `java.util.Set` and `java.util.Map` to real repositories using Spring Data JPA accessing a real database (MySQL 5.7)

Was added in application a docker-compose.yml that contains an image from [mysql/mysql-server:5.7](https://hub.docker.com/r/mysql/mysql-server/). I did not insert the database credentials into application.properties and docker-compose.yml. 

I think that will be necessary modify the fmtok8s-monolith README.md, to help some users to know  how can run this application.

Commit:
Adding flyway, mysql-connector and spring-data-jpa

Commit:
Adding first migration

Commit:
Creating repositories

Commit:
Removing wrong character

Commit:
Changing in memory storage to real relational database

Commit:
Adding docker-compose.yml and modifying application.properties

salaboy commented on docker-compose.yml:
```diff
@@ -0,0 +1,14 @@
+version: '3'
+
```
@mcruzdev the docker-compose file should have both.. the monolith and the database.. :) does that make sense? Do you know how to build the dockerfile from the app? or how to consume it from the external registry which is automatically published?

mcruzdev replied:
Yeas, that make a sense. 
I know how to do this.

I will create a docker image using `docker build`, but will be a local image, is it okay?

Each developer that wants run the monolith will need to generates a image and depending of selected image tag... he will need to modify the docker-compose.yml.

Wich is the best way? Register on docker-hub?

salaboy commented on docker-compose.yml:
```diff
@@ -0,0 +1,14 @@
+    ports:
+    - "3307:3306"
+    environment:
+      MYSQL_ROOT_PASSWORD:
```
@mcruzdev usually it is fine not to include this here in github.. but if we are using the default passwords I would just add them here.. with some comment saying that those are the default ones

salaboy commented on pom.xml:
```diff
@@ -33,11 +33,28 @@
+            <artifactId>spring-boot-starter-data-jpa</artifactId>
+        </dependency>
+
+        <dependency>
```
@mcruzdev this is usually not good, as now our application depends on mysql, and if someone wants to configure it with another database provider they can't. What people usually do is to create a base docker image without any driver and then another layer which adds the jdbc driver of your choice.. can we do that here as well?

mcruzdev replied:
How can i do that ?
I do not know to do .. but i will search about.

salaboy commented on pom.xml:
```diff
@@ -59,6 +76,10 @@
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
@mcruzdev this is duplicated..

salaboy commented on src/main/java/com/salaboy/conferences/site/AgendaController.java:
```diff
@@ -1,57 +1,66 @@
+    private AgendaItemRepository agendaItemRepository;
 
     @PostMapping()
+    @Transactional
```
@mcruzdev interesting.. is this really needed? if so.. why?

salaboy commented on src/main/java/com/salaboy/conferences/site/ProposalService.java:
```diff
@@ -0,0 +1,31 @@
+    public Set<Proposal> getProposals(boolean pending) {
+        if (!pending) {
+
+            return new HashSet<>(repository.findAll());
```
@mcruzdev I would prefer to add a filter going to the database instead of doing it in memory.. can you add that to the ProposalRepository?, it should be just adding an extra method to filter by the proposal status.

salaboy commented on src/main/java/com/salaboy/conferences/site/model/AgendaItem.java:
```diff
@@ -1,17 +1,25 @@
 
-
     public AgendaItem() {
         this.id = UUID.randomUUID().toString();
```
@mcruzdev if you have the generated value, this UUID.random.. is not needed anymore..

mcruzdev replied:
i was doubted about this, thank you ;)

salaboy commented on src/main/java/com/salaboy/conferences/site/repository/ProposalRepository.java:
```diff
@@ -0,0 +1,8 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ProposalRepository extends JpaRepository<Proposal, String> {}
```
@mcruzdev status filter can be added here in the same way as the findAllByDay was added to the AgendaItemRepo

salaboy commented on src/main/resources/application.properties:
```diff
@@ -1 +1,11 @@
 management.endpoints.web.exposure.include=info, health
+
+# datasource
+spring.datasource.url=jdbc:mysql://localhost:3307/{{DATABASE's NAME HERE}}?createDatabaseIfNotExist=true
```
@mcruzdev what is `{{DATABASE's NAME HERE}}` ? I have the feeling that will fail if the user doesn't change that to something.

salaboy reviewed (commented):
@mcruzdev good first approach.. I've added some comments.. this is going to be great when we move all these changes to the microservices approach..

Commit:
Removing unnecessary contructor to generate a UUID

Commit:
Creating a repository method to find our needs

Commit:
Removing duplicated dependency

Commit:
Adding default credentials to docker-compose.yml and application.properties

Commit:
Removing unnecessary annotation @Transactional

Commit:
Adjusting application.properties

Commit:
Removing specific ENGINE and CHARSET

Commit:
Adding h2 database and creating dev spring-profile

Commit:
Testing configuration

salaboy commented on Dockerfile:
```diff
@@ -1,8 +1,18 @@
+FROM maven:3.6.0-jdk-11-slim AS build
```
@mcruzdev why did you changed this? any reason to do that? Multi staged build are well known to cause some issues with pipelines

mcruzdev replied:
I did because was the way that i found to pass a jdbc driver to project disacoupled.

salaboy commented on docker-compose.yml:
```diff
@@ -0,0 +1,31 @@
+      MYSQL_PASSWORD: ${MONOLITH_DATABASE_PASSWORD}
+  monolith_application:
+    container_name: monolith_app
+    image: salaboy/fmtok8s-monolith
```
@mcruzdev this is not the right image, I would use this one `gcr.io/camunda-researchanddevelopment/fmtok8s-monolith` which is the one that is being built by the pipeline

salaboy commented on pom.xml:
```diff
@@ -18,6 +18,8 @@
         <java.version>11</java.version>
         <maven.compiler.source>11</maven.compiler.source>
         <maven.compiler.target>11</maven.compiler.target>
+        <driver.jdbc.group.id>mysql</driver.jdbc.group.id>
```
@mcruzdev this is not actually solving the problem, what I would do for now is to leave it with mysql until we figure out how to add the driver more on the docker side.

salaboy reviewed (commented):
@mcruzdev sorry for the delay, I've added some comments and I am happy to merge if you revert back the Dockerfile which will make the pipeline fail (multistage builds is usually not recommended) . 
Let's keep it with mysql and then we figure that out in a separate PR

Commit:
Removing build step from Dockerfile

Commit:
Removing monolith_app envs for now

mcruzdev commented:
For now, i removed some environments items from docker-compose and added to application-prod.properties.
The items removed are relationed with the MySQL Database like driver class name and dialect.
I removed until we figure out how to added the driver on more docker side.

Commit:
add right image

salaboy reviewed (approved):
@mcruzdev awesome! first PR merged!
